py/src/nupic/tests/encoders/date_test.py

```diff
@@ -0,0 +1,223 @@
+       0,0,0,0,0,0,1,1,1,1,
+       1,0,0,0,0,0,0,0,0,0])
+    expected = seasonExpected + dayOfWeekExpected + weekendExpected + timeOfDayExpected
+
+    print(enc.timeOfDayEncoder.size)
+    print(expected)
+    print(bits.dense.tolist())
+    self.assertEqual(bits.size, 51)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+
+  def testDayOfWeek(self):
+    """ Creating date encoder instance. """
+    # 1 bit for days in a week (x7 days -> 7 bits), no other fields encoded
+    enc = DateEncoder(dayOfWeek=1)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+
+    # Week is MTWTFSS contrary to localtime documentation, Monday = 0 (for
+    # python datetime.datetime.timetuple()
+    dayOfWeekExpected = [0,0,0,1,0,0,0] #Thu
+
+    expected = dayOfWeekExpected
+    self.assertEqual(bits.size, 7)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+  def testSeason(self):
+    """ Creating date encoder instance. """
+    # 3 bits for season (x4 seasons -> 12 bits), no other fields encoded
+    enc = DateEncoder(season=3)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+
+    # Season is aaabbbcccddd (1 bit/month)
+    seasonExpected = [0,0,0,0,0,0,0,0,0,1,1,1]
+
+    expected = seasonExpected
+    self.assertEqual(bits.size, 12)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+  def testWeekend(self):
+    """ Creating date encoder instance. """
+    # 1 bit for weekend (x2 possible values (True=is weekend/False=is not)-> 2 bits), no other fields encoded
+    enc = DateEncoder(weekend=1)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+ 
+    # Not a weekend, so it should be "False"
+    weekendExpected = [1, 0] # "[1,0] = False (category1), [0,1] = True (category2)
+
+    expected = weekendExpected
+    self.assertEqual(bits.size, 2)
+    self.assertEqual(expected, bits.dense.tolist())
+
+
+  def testTime(self):
+    """ Creating date encoder instance. """
+    # 3 bits for season (x4 seasons -> 12 bits), no other fields encoded
+    enc = DateEncoder(timeOfDay=5)
+    # In the middle of fall, Thursday, not a weekend, afternoon - 4th Nov,
+    # 2010, 14:55
+    d = datetime.datetime(2010, 11, 4, 14, 55)
+    # d = datetime.datetime(2010, 11, 1, 8, 55) # DEBUG
+    bits = enc.encode(d)
+
+    # Time of day has radius of 4 hours and w of 5,
+    # so each bit = 240/5min = 48min. 
+    # 14:55 is minute 14*60 + 55 = 895; 895/48 = cetral bit= bit 18.6 ->18 #FIXME we should round this, so becomes bit19
```

Comment:
this 18.6 should round up to 19. Is that the behavior of our code (should be fixed), or just a mistake in the comment and expected data?

Reply:
@ctrl-z-9000-times ?